API/Tensorflow/native_libs/src/c_wrappers/operations.cpp

```diff
@@ -335,21 +335,98 @@ void** eval_graph_with_placeholders(GraphSession *graph,
     };
 }
 
+namespace {
+    double to_double(std::shared_ptr<Tensor> tensor, int64_t index) {
+        switch(tensor->getType()) {
+            case TF_FLOAT:
+                return static_cast<double>(tensor->at<TF_FLOAT>(index));
+            case TF_DOUBLE:
+                return tensor->at<TF_DOUBLE>(index);
+            default:
+                throw std::runtime_error("Loss type not supported by loss history");
+        }
+    }
+
+    double iterate_over_inputs(GraphSession* graph, const char** ph_names, Tensor** ph_values, uint32_t ph_count,
+            std::shared_ptr<State>& state, uint32_t start_index, uint32_t end_index, bool apply_side_effects = true) {
+        std::map<std::string, std::shared_ptr<Tensor>> substitutions;
+        double running_loss = 0;
+        for (uint32_t input_index = start_index; input_index < end_index; ++input_index) {
+            for (size_t ph = 0; ph < ph_count; ++ph) {
+                substitutions.emplace(std::string(ph_names[ph]),
+                                      LifetimeManager::instance().accessOwned(ph_values[input_index * ph_count + ph]));
+            }
+            std::shared_ptr<EvaluationResult> eval_result = graph->eval(substitutions, state, apply_side_effects);
+            running_loss += to_double(eval_result->outputs[0], 0);
```

Comment:
This will throw an exception when loss is not convertible to double as in `to_double`.
I don't think making it impossible at all to train models with other losses is a good idea.
I understand it's rather unlikely to have integer losses, but it's certainly possible so maybe it would be good to allow for this. One way might be to catch the exception and just ignore loss history?